source/assets/js/playground/autocomplete.ts

```diff
@@ -0,0 +1,241 @@
+import {
+  CompletionContext,
+  CompletionResult,
+  CompletionSource,
+} from '@codemirror/autocomplete';
+import {sassCompletionSource} from '@codemirror/lang-sass';
+import {syntaxTree} from '@codemirror/language';
+import {EditorState} from '@codemirror/state';
+import moduleMembers from './module-members';
+
+// validFor identifier, from @codemirror/lang-css
+const identifier = /^(\w[\w-]*|-\w[\w-]*|)$/;
+
+// Sass-specific at rules only. CSS at rules should be added to `@codemirror/lang-css`.
+const atRuleKeywords = [
+  'use',
+  'forward',
+  'import',
+  'mixin',
+  'include',
+  'function',
+  'extend',
+  'error',
+  'warn',
+  'debug',
+  'at-root',
+  'if',
+  'else',
+  'each',
+  'for',
+  'while',
+];
+
+const atRuleOptions = Object.freeze(
+  atRuleKeywords.map(keyword => ({
+    label: `@${keyword} `,
+    type: 'keyword',
+    validFor: identifier,
+  }))
+);
+
+// Completions for Sass at rules
+function atRuleCompletion(context: CompletionContext): CompletionResult | null {
+  const atRule = context.matchBefore(/@\w*/);
+  if (!atRule) return null;
+  if (atRule.from === atRule.to && !context.explicit) return null;
+  return {
+    from: atRule.from,
+    to: atRule.to,
+    options: atRuleOptions,
+    validFor: identifier,
+  };
+}
+
+type CompletionInfo = {
+  name: string;
+  description?: string;
+};
+
+type CompletionModule = CompletionInfo & {
+  functions?: CompletionInfo[];
+  variables?: CompletionInfo[];
+};
+
+const moduleNames = moduleMembers.map(mod => mod.name);
```

Comment:
This value reads weird—you're iterating over module _members_, but the variable name suggests that it's a _module_ rather than a member of one. I think this suggests that the name "module-members.ts" isn't really accurate. Maybe call it "module-metadata.ts" instead?